Allow a single mouse drag across the EQ to set all values

chriscareycode opened the pull request:
In WinAMP you could mouse down on the first EQ VerticalSlider and drag the mouse across all of the VerticalSliders to set them all in one swipe. This PR brings the same capability to WebAMP.

https://user-images.githubusercontent.com/1002638/177918378-2615b514-f093-417a-85aa-0ad902d3c3e0.mov

captbaritone commented on packages/webamp/js/components/WinampButton.tsx:
```diff
@@ -71,12 +82,19 @@ export default function WinampButton(props: Props): JSX.Element {
   );
 
+  const onMouseEnter = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
+    if (e.buttons === 1) {
```
Oh that's clever! I didn't realize you could check that on mouse enter.

captbaritone commented on packages/webamp/js/skinSelectors.ts:
```diff
@@ -156,7 +156,7 @@ export const imageSelectors: Selectors = {
   EQ_SLIDER_BACKGROUND: [".band"],
   EQ_SLIDER_THUMB: [".band .slider-handle"],
-  // But the "active" pseudo selector on the parent, since clicking
+  // Put the "active" pseudo selector on the parent, since clicking
```
Thanks!

captbaritone commented on packages/webamp/js/components/EqualizerWindow/Band.tsx:
```diff
@@ -42,24 +42,26 @@ export default function Band({ id, onChange, band }: Props) {
   }, [value]);
   const focusBand = useActionCreator(Actions.focusBand);
-  const usetFocus = useActionCreator(Actions.unsetFocus);
+  const unsetFocus = useActionCreator(Actions.unsetFocus);
```
Thanks!

captbaritone commented on packages/webamp/js/components/VerticalSlider.tsx:
```diff
@@ -99,9 +101,16 @@ export default function VerticalSlider({
       document.addEventListener("mouseup", handleMouseUp);
+      if (type === "equalizer") {
+        // If the mouse leaves the Slider area, trigger a MouseUp
+        sliderNode.addEventListener("mouseleave", handleMouseUp);
```
Since this is now triggered on more than just mouse up, can we rename `handleMouseUp` to `handleRelease` or something like that?

captbaritone commented on packages/webamp/js/components/WinampButton.tsx:
```diff
@@ -47,8 +52,14 @@ export default function WinampButton(props: Props): JSX.Element {
       }
       document.addEventListener("mouseup", onUp);
+      if (type === "equalizer") {
+        e.target.addEventListener("mouseleave", onUp);
```
Similar to above, can we rename `onUp` to `onRelease`?

captbaritone reviewed (changes requested):
Thanks for this! This has been one of the gaps in Webamp that has bothered me for a long time.

A few requests:

1. Some naming nits (see inline comments)
2. Rather than `type="equalizer"` (which tightly couples the generic components to the implicit behavior of the equalizer) can we do a boolean prop called `requireClicksOriginateLocally` (or similar) that defaults to `true`? This is a behavior that's actually common to many buttons in Winamp, so we'll want to enable this more broadly as a followup.
3. Can you also handle touch events so that this behavior works on mobile as well?

chriscareycode commented:
1. Done

2. When switching the prop to requireClicksOriginateLocally, the WinampButton is giving the warning:

Warning: React does not recognize the `requireClicksOriginateLocally` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `requireclicksoriginatelocally` instead. If you accidentally passed it from a parent component, remove it from the DOM element.

If we lowercase the parameter and use it on the JSX.Element component, we are going to need the parameter to be a string and not boolean for it to be happy. So, we probably want to refactor WinampButton to return a standard React functional component and not return a JSX.Element. I'm not sure if you had specific reasons for doing this component in this way. What do you suggest I should do here?

3. Adding support for touch is going to be a bit more tricky.

touchmove event target is always the same element which started the touch, so if we touch on band 1000 and move over other EQ bands, the event only continues to fire on the original element.

W3C added and subsequently removed touchenter and touchleave events, which would have made this nice and clean, but we do not have those available.
https://www.w3.org/TR/2011/WD-touch-events-20110505/ old
https://www.w3.org/TR/touch-events/#list-of-touchevent-types current

We are probably going to need to track touchmove on the parent component (containing all the EQ bands), and if the x,y coordinates match the child element, to activate change in the children. I’m up for the challenge to give it a try, but it will take a bit longer to get through this and get the code clean and efficient. I made some initial efforts but things started to get a bit hairy. I’m thinking a follow up PR might be in order unless you feel strongly to wait for both to roll this out.

chriscareycode commented:
Got touch working with technique 3, with most of the code living in EqualizerWindow/index.tsx. Going to sit and stare at it for a bit.

chriscareycode commented:
Related to Touch: I noticed a bug where if you touch the EQ, then the next subsequent touch on button like stop, play, will not work the first time you touch it. To repro: Touch Play. Touch Stop. Touch the EQ. Touch Play. It will not play, until you touch it a second time. I tested this on winamp.org and seeing the bug there so it appears to be unrelated to this PR.

captbaritone commented on packages/webamp/js/components/EqualizerWindow/index.tsx:
```diff
@@ -35,6 +37,124 @@ const EqualizerWindow = () => {
+   * @param clientX 
+   * @returns Band
+   */
+  const getBandFromClientX = (clientX: number) => {
```
I don't think this approach (where we bind at the top level window, and then calculate which element the mouse is referring to) is correct. For one it looks very brittle, but for two, I don't think it matches the behavior of Winamp. In Winamp it looks like the you are always dragging the last active slider, not just the one that is closest to your cursor on the x axis:

https://user-images.githubusercontent.com/162735/176981011-0064735a-d25b-4d82-ba50-6c4297c3c2f8.mov

chriscareycode replied:
Yeah I'm with you. It was quite a hack. I'm tearing all that code out in favor of PointerEvents. In my local dev it's working well and the new code is much more clean. With the video you posted it also shows some other behavior that would need some modification to get in sync with. Like, we don't want to "let go" of the band when you leave the verticalslider area, but only when you release the mouse/touch, or enter another verticalslider. So some changes to how "mouseleave"/"pointerleave" was being utilized.

chriscareycode replied:
This is fixed and now behaves the same as the video.

captbaritone commented on packages/webamp/js/components/EqualizerWindow/index.tsx:
```diff
@@ -35,6 +37,124 @@ const EqualizerWindow = () => {
+   * Set ACTIVE_CLASSNAME on the current band (to set the selected button state)
+   */
+  const setWinampActiveClass = (band: BandType) => {
+    document.querySelector(`#band-${band}`)?.classList.add(ACTIVE_CLASSNAME);
```
Let's try to find a solution that uses React rendering to set the DOM state rather than manually manipulating the DOM. If manual DOM manipulation is absolutely required (I don't think it will be), it should happen within the component that rendered the DOM element.

chriscareycode replied:
This is fixed in local dev and this hack is no longer necessary

captbaritone commented on packages/webamp/js/components/EqualizerWindow/index.tsx:
```diff
@@ -35,6 +37,124 @@ const EqualizerWindow = () => {
+    const bandElements = document.querySelectorAll(`.band`);
+    bandElements.forEach(bandEl => {
+      bandEl?.classList.remove(ACTIVE_CLASSNAME);
+    });
```
See above, let's avoid this pattern of finding DOM elements and mutating them.

chriscareycode replied:
This is fixed in local dev and this hack is no longer necessary

chriscareycode commented:
Ready for review. Github says "captbaritone requested changes" but I'm not sure how to ACK that or let it know the changes are done.

captbaritone commented:
Looking better! One thing that's still not right though: In real Winamp, clicking outside the eq window and then dragging across the sliders does not cause the sliders to be moved. Also, clicking _inside_ the eq window and the dragging over the preamp causes us to grab the preamp slider but we can't then glide into the band sliders.

I think I'd be happy with an abstraction that gave us a separate React component that defined an area that would capture these initial clicks. So, any click that happened in this new component or one of its children would trigger the behavior you have now, but if the click originated outside of an instance of this new component, the drags would be ignored.

chriscareycode commented on packages/webamp/js/components/WinampButton.tsx:
```diff
@@ -28,56 +32,62 @@ type Props = DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>;
+  };
 
+  // Remove the custom prop "requireClicksOriginateLocally" before passing the HTML props to the child div
+  const htmlProps = {...props};
+  delete htmlProps.requireClicksOriginateLocally;
```
Neat trick. Can't do this here though. requireClicksOriginateLocally is already declared up in the function params. This will redeclare the same variable as part of the destructuring.

chriscareycode replied:
Thx. Updated in PR

chriscareycode commented on packages/webamp/js/components/WinampButton.tsx:
```diff
@@ -28,56 +32,62 @@ type Props = DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>;
   const [active, setActive] = useState(false);
-  const originalOnMouseDown = props.onMouseDown;
+  const originalOnPointerDown = props.onPointerDown;
+  const { requireClicksOriginateLocally } = props;
```
Thx. Updated in PR

chriscareycode commented on packages/webamp/js/components/VerticalSlider.tsx:
```diff
@@ -31,12 +31,13 @@ export default function VerticalSlider({
   onBeforeChange,
   onChange,
   onAfterChange,
+  requireClicksOriginateLocally,
```
Thx. Updated in PR

chriscareycode commented:
Getting closer. Fixed the bug where clicking outside the EQ area and dragging across the EQ will affect the sliders. Also fixed the issue that dragging across the EQ will affect the preamp. Now the drag will only work if the click originates in the EQ area by adding a clickOriginatedInEq state.

Still have a bug where if you release way off to the side, off the WinAmp area, the clickOriginatedInEq state does not get reset. Thinking about a clean solution for that.

chriscareycode commented:
Working better now. Tested with mouse, pen, and touch.

captbaritone commented:
Thanks for all your work on this, and most of all your patience! I've been busy with other stuff so I haven't been able to muster the focus to give timely review. Sorry for that. There are a few details to address here, but rather than go back for another round, I'll go ahead and merge this and followup with the fixes when I find time. (I'll mention you so you can see, in case you are curious or have feedback for me on why you chose to do it differently.

This is such a great detail to have right on Webamp. I really appreciate you persevering and getting it added!
